### PR TITLE
Fixes #28519: Hash and run database seeds on startup 

### PR DESCRIPTION
**Pull request by ehelms**
This is part of moving to a model where packaging can remove this file to indicate seeding needed and the installer can trigger if the file does not exist.

### CODE REVIEW

File: lib/tasks/config.rake
Comment (ekohl) [inline lib/tasks/config.rake]: This probably needs to be configurable since this will fail in development setups. I think we have a directory relative to the app_root with symlinks in place in production.

Comment (timogoebel) [reply lib/tasks/config.rake]: How is this going to work on a multi host setup?

**Review by timogoebel — CHANGES_REQUESTED** [review]

File: lib/tasks/config.rake
Comment (ehelms) [reply lib/tasks/config.rake]: Not very well I suppose. But if you are using the puppet modules on each host and updating packages on each host seems that they would all end up in the right state to database migrate and database seed. This is mostly targeting an attempt to simplify our packages, and keep the puppet modules idempotent.

I'm open to other ideas. Seeding is hard because they are idempotent in and of themselves and therefore there is no method to detect that they have ran like migrations.

File: lib/tasks/config.rake
Comment (ekohl) [inline lib/tasks/config.rake]: To be clear, I think we should *never* hardcode a path. We already have code that makes sure `/var/lib/foreman/db` is a symlink to `/usr/share/foreman/db`. That means I think we should use `File.join(Rails.root, 'db', 'seeded')` because it'll work in both production and development. Note 100% sure about permissions though.

https://github.com/theforeman/foreman-packaging/blob/51812c4a35e4b0e010022777371a80d8939921db/packages/foreman/foreman/foreman.spec#L823

Comment (timogoebel) [reply lib/tasks/config.rake]: I think we should _never_ save any state in files. This should be saved in the database.

Comment (ehelms) [reply lib/tasks/config.rake]: Do you have any suggestions on how to solve this given:

 1) Puppet module doesn't want to run db:seed everytime
 2) RPMs shouldn't rely on the needing the database to be running to get updated

Comment (ekohl) [reply lib/tasks/config.rake]: Currently we have a setting that it was seeded, but we have no way to indicate a seed is needed again.

With database migrations, we store all the files we ran but seeds should be idempotent and files are ran again. That means you need to store the time. You could use `max(seed.mtime for seed in seeds)` and store that. It probably means that when you add a new node to the cluster you'll seed again.

In these cases I always look at what Django has done and they don't have seeds but rather data migrations which are handled in the same way as schema migrations.

Taking a step back, we could treat seeds the same way since they're in `seeds.d`. You never modify a seed but only add files. If you need to alter, you can rename it. However, that might be diverting from vanilla rails too much. Some seeds also depend on the order.

File: lib/tasks/config.rake
Comment (ekohl) [reply lib/tasks/config.rake]: Another issue with storing it in the database is that multiple hosts might both be looking at the same value and decide they need to concurrently perform an action. That means you need to create a lock as well.

File: lib/tasks/config.rake
Comment (timogoebel) [reply lib/tasks/config.rake]: > 1. Puppet module doesn't want to run db:seed everytime

Basically, I'd extend `db/seeds.rb` to save the following information in a database table: `rails_engine_name`, `seed_filename`, `seed_file_sha256` and `migration_date`. If the sha256 changes or the entry does not exist, we need to run the seed.

> 2\. RPMs shouldn't rely on the needing the database to be running to get updated

I'd just run the seed logic every time the rails app starts. That needs the database running anyways. Neither the rpm, the admin or puppet needs to do anything here.

Comment (ekohl) [reply lib/tasks/config.rake]: > > 1. Puppet module doesn't want to run db:seed everytime
> 
> Basically, I'd extend `db/seeds.rb` to save the following information in a database table: `rails_engine_name`, `seed_filename`, `seed_file_sha256` and `migration_date`. If the sha256 changes or the entry does not exist, we need to run the seed.

I like the idea, but would it also work with cases like:

https://github.com/theforeman/foreman/blob/develop/db/seeds.d/020-permissions_list.rb
https://github.com/theforeman/foreman/blob/a9b03964bd5cb3a0f039af8b6f97a4d8686c289c/db/seeds.d/020-roles_list.rb#L1
https://github.com/theforeman/foreman/blob/a9b03964bd5cb3a0f039af8b6f97a4d8686c289c/db/seeds.d/030-permissions.rb#L1-L5

I believe this is done so plugins can extend the permissions list. This probably works if you're very strict that seeds can only require other seed files.

> > 2. RPMs shouldn't rely on the needing the database to be running to get updated
> 
> I'd just run the seed logic every time the rails app starts. That needs the database running anyways. Neither the rpm, the admin or puppet needs to do anything here.

This feels like overkill. However, if you you can determine if you need to run by only querying a single table and `map(sha256, seed.d/*)` it may be cheap enough.

Odd edge case, but calculating the bcrypt cost might be misleading if the app is starting up because it'll likely saturate the CPU core with unrelated work. https://github.com/theforeman/foreman/blob/a9b03964bd5cb3a0f039af8b6f97a4d8686c289c/db/seeds.d/010-calibrate_bcrypt_cost.rb#L1-L2

I'd like to see this encapsulated and exposed in a rake task as well.

What do you think about trucating the seed table to make sure you purge all old, possibly non-existing seed entries when seeding?

Comment (timogoebel) [reply lib/tasks/config.rake]: > I believe this is done so plugins can extend the permissions list. This probably works if you're very strict that seeds can only require other seed files.

I believe there is a proper plugin API for this. Templates might be an issue, but we can find a solution for this. Maybe generating the sha over the templates as well.

> This feels like overkill. However, if you you can determine if you need to run by only querying a single table and map(sha256, seed.d/*) it may be cheap enough.

Sure, that should work as well.

> What do you think about trucating the seed table to make sure you purge all old, possibly non-existing seed entries when seeding?

We could remove all entries where the seed files do not exist anymore.

Comment (ehelms) [reply lib/tasks/config.rake]: Sounds a bit like we are implementing database migration tracking instead of just using database migrations in Rails. Is there enough of a difference to support both?

Comment (timogoebel) [reply lib/tasks/config.rake]: Of course you can use the simple timestamp approach and just store that value in a (hidden) setting. If max(filemtimes) > the saved timestamp, just run the seeds.

Comment (ehelms) [reply lib/tasks/config.rake]: Either of y'all mind doing a recap of where we landed with the implementation idea?

Comment (timogoebel) [reply lib/tasks/config.rake]: I'm not sure if we actually reached a conclusion.
@ehelms: Do you mind describing your initial idea? I'm still not sure what you had initially in mind.

Comment (ehelms) [reply lib/tasks/config.rake]: The initial idea was:

 1) When database seeding occurs, create a file on disk
 2) puppet module will run db:seed if that file does not exist
 3) RPM packages or otherwise would remove this file on update to trigger a database seed by the puppet module

This prevents the RPMs from touching the database and the puppet module from running database seed everytime and triggering a false positive change event within puppet.

I agree it's not the nicest way but it achieved the end result. There is arguably a larger problem in how we overuse database seeding. Their intent as I understand is to load initial data into a database and in some cases we apply this pattern to upgrades. If we had a hard stance on initial data we'd run seeding just once and never again.

Comment (timogoebel) [reply lib/tasks/config.rake]: The more I think about it is, that we should run the seeds on the application start. That would force us to keep seeds fast and figure out a way to run them just once.

Comment (ehelms) [reply lib/tasks/config.rake]: I wonder if that is a change we could enforce as 2.0? Any edge cases would then force us to cover them either through a plugin API or use of database migrations. @tbrisker for your thoughts

Comment (tbrisker) [reply lib/tasks/config.rake]: It feels to me like using a file as a signal between rpm and puppet is not the best solution for this, for reasons mentioned before. 
Perhaps, we could calculate a hash that includes all the seed files that ran in https://github.com/theforeman/foreman/blob/develop/db/seeds.rb#L26 (where we already have the plugin seeds also) and save that in the database, than at startup calculate the hash again and if it doesn't match automatically rerun the seeds. This will catch any changes to any seed or configuration, and actually remove the "rake db:seed" step from the installation (as it will automatically execute on the first startup since the initial hash will be blank). This will also cover changes to e.g. the permissions list.
Regarding using seeds to modify data in existing installation - I agree that should be done in migrations generally. However, I think in some cases a migration won't work for new installations since some data won't be seeded yet so it won't be modified (e.g. assigning resources to the default taxonomies). I'm not sure how we can enforce making sure we have a migration that matches any change in the seed though, other than by the reviewers asking for it.

**Comment by ehelms** [comment]
This has been updated based on @tbrisker idea. Have a look at let me know further thoughts. I would prefer to drop the settings in a separate PR after this concept is merged.

File: config/initializers/seeds.rb
Comment (timogoebel) [inline config/initializers/seeds.rb]: I might be nitpicking here, but could this also be a (hidden) setting?

**Review by timogoebel — COMMENTED** [review]
Nice, this is roughly what I had in mind. Thanks a bunch, @ehelms!

File: config/initializers/seeds.rb
Comment (timogoebel) [reply config/initializers/seeds.rb]: ... that would make this easier. If you would save the filenames with the individual hashes, you could speed this up a little. And we’d probably have a good example for over engineered code.

Comment (ehelms) [reply config/initializers/seeds.rb]: I need a bit more help understanding here. Hidden setting? Is that simply avoiding a new table to use the existing Settings table to store the hash and keep it from the user? If I went that route, how would I store every individual file and hash? As a JSON blob?

On my devel setup (with Katello and foreman-tasks), the hash calculation and database lookup take 0.013356433 seconds.

Comment (timogoebel) [reply config/initializers/seeds.rb]: If you just store a single hash, a setting might be a cheap way to store this data and avoids the extra table. If you want to store an individual hash per file, the extra model is the better fit imho.
Optimizing this for speed would be easy to do, but it’s probably unnecessary.

Comment (ehelms) [reply config/initializers/seeds.rb]: I'm not opposed to that route. @tbrisker @ekohl  your thoughts when you have a chance

Comment (ekohl) [reply config/initializers/seeds.rb]: As always, @timogoebel has a great idea. Do we have a way to hide settings in the UI?

File: config/initializers/seeds.rb
Comment (ekohl) [inline config/initializers/seeds.rb]: I think the alternative is:
```ruby
hash = Digest::SHA256.new
foreman_seeds.each { |seed| hash.file(seed) }
new_hash = hash.base64digest
```

Not sure it matters

File: config/initializers/seeds.rb
Comment (ekohl) [inline config/initializers/seeds.rb]:
```suggestion
    foreman_seeds = foreman_seeds.sort_by { |seed| seed.split('/').last }
```

File: config/initializers/seeds.rb
Comment (ekohl) [inline config/initializers/seeds.rb]: Does root have a trailing slash? Would `File.join` be better? Same below for the glob

**Review by ekohl — COMMENTED** [review]
This still allows explicitly calling `db:seed`. Does that now seed twice? Should we enhance that job to store the seed as well?

File: config/initializers/seeds.rb
Comment (tbrisker) [reply config/initializers/seeds.rb]: is it be worth the effort to save each file's hash? since some files import others and seeding should be idempotent, we'd want to rerun seeds anyways if the combined hash changes. 0.01 seconds at startup to check if we need to run the seeds doesn't sound like much.

File: config/initializers/seeds.rb
Comment (tbrisker) [inline config/initializers/seeds.rb:3]: This feels like we're duplicating the logic from seed.rb? perhaps the hash checking should move there and the initializer should just trigger the rake task? or the logic extracted to a service that both call?

**Comment by ehelms** [comment]
I lean towards `rake db:seed` should still do the more normal Rails behavior of running all seed files. This gives users the opportunity to force run the seeds and keeps the standard Rails behavior in tact. I see this new method, running on startup, as a way for us to control our upgrade and install behavior through our various tools (puppet, installer, etc.)

File: config/initializers/seeds.rb
Comment (ehelms) [reply config/initializers/seeds.rb:3]: Absolutely, I should have mentioned that as a to-do still. My initial goal was get a feel for the implementation which sounds like folks are good with.

File: config/initializers/seeds.rb
Comment (ehelms) [reply config/initializers/seeds.rb]: Possibly. This works today as this code is largely borrowed from `db/seeds.rb` to test the implementation.

**Comment by ekohl** [comment]
> I lean towards `rake db:seed` should still do the more normal Rails behavior of running all seed files. This gives users the opportunity to force run the seeds and keeps the standard Rails behavior in tact. I see this new method, running on startup, as a way for us to control our upgrade and install behavior through our various tools (puppet, installer, etc.)

I meant that `db:seed` should store the hash in the same way as start up does to prevent duplicate seeding on start up. It should still seed regardless of the previous seeding state.

**Comment by tbrisker** [comment]
> I lean towards `rake db:seed` should still do the more normal Rails behavior of running all seed files. This gives users the opportunity to force run the seeds and keeps the standard Rails behavior in tact. I see this new method, running on startup, as a way for us to control our upgrade and install behavior through our various tools (puppet, installer, etc.)

I tend to agree. we could use the same service for both, and change the condition for rerun - i.e. something like `if Foreman.in_rake?('db:seed') || hash_changed?`

**Comment by ehelms** [comment]
Is hidden settings a thing? I'm struggling to find a reference

**Comment by tbrisker** [comment]
> Is hidden settings a thing? I'm struggling to find a reference

I don't think so. I wonder if we should create some table for internal key value store for things like this. Possibly there are other settings (db_pending_migration and db_pending_seed come to mind, but there may be others) that were pushed into settings because we had no good place to put them but should be moved to it as well.

**Comment by ehelms** [comment]
The two examples, db_pending_migration and db_pending_seed, will be removed after this work is completed. Are there other examples that would warrant a separate key value table? Even if not, I could go that route here to move this PR along and open up the possibility for the future. 

I admit, bikeshedding wise, I have no idea what to call said table and model.

**Comment by ehelms** [comment]
This is updated to centralize the code into a class and use it in both the initializer and traditional seed code. This still uses the dedicated database table while we hash out (no pun intended) where we want to store this information.

File: config/initializers/seeds.rb
Comment (tbrisker) [inline config/initializers/seeds.rb]: I would move this into the `.execute` step, so the hash is still recorded even if seeds are executed manually

File: config/initializers/seeds.rb
Comment (tbrisker) [inline config/initializers/seeds.rb]: Pretty sure you can't return here since it's not a function. 
We should probably skip this in any rake task, not just db ones, as well as in test env. We only want to seed on actual app startup.

File: app/models/database_seed.rb
Comment (tbrisker) [inline app/models/database_seed.rb]: I still feel it's weird to have a "singleton table". I would prefer something more generic such as `ForemanInternal` or some such that is a key value store that could be used for other internal details users shouldn't care much about. I suppose if we go over all settings we would probably find a few others that could be moved to it and were only there because we have no other good place to store them persistently (and maybe also things that are currently hard coded or saved in various config yamls).

File: app/services/foreman_seeder.rb
Comment (tbrisker) [reply app/services/foreman_seeder.rb]: the hound is right :)

File: config/initializers/seeds.rb
Comment (ehelms) [reply config/initializers/seeds.rb]: Well, it seems to work so I am guessing you can :)

**Comment by ehelms** [comment]
Updated to move to a ForemanInternals table as a generic option for storage. The database fields are 'key' and 'value' to indicate it's simplistic key/value storage nature.

File: config/initializers/seeds.rb
Comment (tbrisker) [reply config/initializers/seeds.rb]: That's interesting, looks like you can indeed. learned something new.

File: config/initializers/seeds.rb
Comment (tbrisker) [reply config/initializers/seeds.rb]: I think this is still valid - the hash should be updated if we run the seeds manually as well

File: config/initializers/seeds.rb
Comment (tbrisker) [inline config/initializers/seeds.rb]: Should we also return in test env just to make sure it doesn't get seeded by mistake somehow?

File: config/initializers/seeds.rb
Comment (ehelms) [reply config/initializers/seeds.rb]: I was resisting this because its a minor savings but adds complexity to the ForemanSeeder class as it would now be tied to knowledge of database entities. I can go that route still just wanted to clarify my reasoning.

**Comment by ehelms** [comment]
@timogoebel where are you thinking the compact needs to happen?

**Comment by timogoebel** [comment]
> @timogoebel where are you thinking the compact needs to happen?

@ehelms:
```ruby
    rescue NameError => e
      Foreman::Logging.exception("Failed to register plugin #{name}", e)
    end.flatten.compact
```
When a `NameError` is rescued, ....

File: db/seeds.rb
Comment (tbrisker) [inline db/seeds.rb:14]: I you don't want to do the updating inside the seeder service, how about adding 
```ruby
ForemanInternal.find_or_create_by(key: 'database_seed').update_attribute(:value, seeder.hash) 
```
here?

**Comment by tbrisker** [comment]
Looks like test failures are related

**Comment by ehelms** [comment]
[test katello]

**Comment by ehelms** [comment]
[test katello]

File: db/seeds.rb
Comment (ehelms) [reply db/seeds.rb:14]: Done... I just put it in the foreman_seeder service in the end and simplified things.

**Review by tbrisker — CHANGES_REQUESTED** [review]
There's still one failing test, other than that LGTM!

**Comment by tbrisker** [comment]
[test katello]

**Review by tbrisker — DISMISSED** [review]
LGTM! i'll let the butler finish and give a chance for @ekohl and @timogoebel to also ack before merging.

**Review by timogoebel — DISMISSED** [review]

File: app/services/foreman_seeder.rb
Comment (ekohl) [inline app/services/foreman_seeder.rb:21]: This: this isn't really registration.

File: app/services/foreman_seeder.rb
Comment (ekohl) [inline app/services/foreman_seeder.rb:56]: If it is created, does it create it with nil as value?

File: config/initializers/seeds.rb
Comment (ekohl) [inline config/initializers/seeds.rb]: Is it better to use logging here or can't you rely on that in initializers?

**Review by ekohl — DISMISSED** [review]
Some questions that partly are for my understanding of Rails.

File: app/services/foreman_seeder.rb
Comment (ehelms) [reply app/services/foreman_seeder.rb:21]: The wording aint the best per say

File: app/services/foreman_seeder.rb
Comment (ehelms) [reply app/services/foreman_seeder.rb:56]: It does

File: app/services/foreman_seeder.rb
Comment (ekohl) [reply app/services/foreman_seeder.rb:21]: I later saw that it was just moved from somewhere else. Might be a good chance to improve but also fine to keep it.

**Comment by ehelms** [comment]
[test katello]

**Comment by ehelms** [comment]
[test katello]

**Review by ekohl — APPROVED** [review]

**Comment by ehelms** [comment]
I made a few updates to switch to using the logger instead of puts. Once test finish up @tbrisker I'll leave this to your discretion.

**Review by tbrisker — APPROVED** [review]
Thanks @ehelms !

**Comment by timogoebel** [comment]
Thanks a bunch, @ehelms.

Do we need to update the manual so users know they don't need (but can) to run `foreman-rake db:seed` anymore?

**Pull request merged**
